Cache results in account witness domain

chimp1984 opened the pull request.

chimp1984 committed:
Add accountAgeWitnessCache

The accountAgeWitnessMap is very large (70k items) and
access is a bit expensive. We usually only access less
than 100 items, those who have offers online. So we
use a cache for a fast lookup and only if
not found there we use the accountAgeWitnessMap and
put then the new item into our cache.

chimp1984 committed:
Add getSignedWitnessSetCache

The getSignedWitnessSet is called very often and is a bit
expensive. We cache the result in that map but we
remove the cache entry if we get a matching SignedWitness
added to the signedWitnessMap.

chimp1984 committed:
Fix failing test (remove did operate on map
directly and we did not remove the item from the cache)

- Add getSignedWitnessMapValues method for access to signedWitnessMap values
- Remove Getter for signedWitnessMap
- Add removeSignedWitness method (used in test only)
- Use addToMap in test instead of direct access to map
- Remove getSignedWitnessSetCache entry matching
signedWitness.getHashAsByteArray() as well.
See comment. @sqrrm: Can you cross check?

chimp1984 commented:
Here are results from performance tests (based on live network data):

I used a loop for 5 iterations over each offer. For findWitness 10000 iterations.

Version of that PR:
652 offers. 6520000 calls of accountAgeWitnessService.findWitness(offer) took 783 ms  (84% to master)
652 offers. 3260 calls of accountAgeWitnessService.getSignState(accountAgeWitness) took 2063 ms (71% to master)
652 offers. 3260 calls of accountAgeWitnessService.getWitnessSignAge(accountAgeWitness, now) took 4094 ms (78 %to master)
652 offers. 3260 calls of accountAgeWitnessService.getAccountAge(accountAgeWitness, now) took 4094 ms (78% to master)
652 offers. 3260 calls of signedWitnessService.getSignedWitnessSet(accountAgeWitness) took 4095 ms (74% to master) 
652 offers. 3260 calls of signedWitnessService.getWitnessDateList(accountAgeWitness) took 4096 ms (70% to master) 

Master version:
653 offers. 6530000 calls of accountAgeWitnessService.findWitness(offer) took 924 ms 
653 offers. 3265 calls of accountAgeWitnessService.getSignState(accountAgeWitness) took 2873 ms 
653 offers. 3265 calls of accountAgeWitnessService.getWitnessSignAge(accountAgeWitness, now) took 5226 ms 
653 offers. 3265 calls of accountAgeWitnessService.getAccountAge(accountAgeWitness, now) took 5226 ms 
653 offers. 3265 calls of signedWitnessService.getSignedWitnessSet(accountAgeWitness) took 5501 ms 
653 offers. 3265 calls of signedWitnessService.getWitnessDateList(accountAgeWitness) took 5788 ms 

I expected that the findWitness call is more expensive but seems thats not the bottleneck. So the optimization with `accountAgeWitnessCache` is prob. not making much difference. Though that is a pretty trivial one.
The optimization with `getSignedWitnessSetCache` seems to have a bit more impact.

I think it requires further investigations where the bottlenecks are in the tested methods.

Overall that PR brings still about a 20-30% improvement.
There are about 3000-5000 signedWitnessService.getSignedWitnessSet calls with normal user activity after a few minutes which accumulates to about 6-10 sec. cpu time in all.

As its a bit risky domain I have no strong opinion about to merge that PR or close it and delay for a later more broad investigation and improvement. Up to maintainers....

chimp1984 committed:
Use ConcurrentHashMap for cache.

As the normal maps are not using a ConcurrentHashMap it is
likely unnecessary as well that I am not aware of a
multi-threaded access.
But as it does not show any difference in performance
it is likely the bit more safe option.

chimp1984 commented:
Seems some Travis issue again: "Could not resolve com.google.api.grpc:proto-google-common-protos:1.12.0." Will force push...

chimp1984 commented:
Travis has issues again... happens quite often recently ;-( Will force push...

sqrrm commented on core/src/main/java/bisq/core/account/sign/SignedWitnessService.java:
```diff
@@ -75,10 +76,15 @@
+
+    // The getSignedWitnessSet is called very often and is a bit expensive. We cache the result in that map but we
+    // remove the cache entry if we get a matching SignedWitness added to the signedWitnessMap.
+    private final Map<P2PDataStorage.ByteArray, Set<SignedWitness>> getSignedWitnessSetCache = new ConcurrentHashMap<>();
```
The name is a bit off for a map, just `signedWitnessSetCache` would be more natural.

chimp1984 replied:
Agree...

sqrrm commented on core/src/main/java/bisq/core/account/sign/SignedWitnessService.java:
```diff
@@ -487,8 +509,18 @@ private boolean verifyDate(SignedWitness signedWitness, long childSignedWitnessD
+        // Not sure if that is needed as well, tests did succeed in both cases, but seems to be more safe to remove
+        // potential entries with hash as well. A removed item in getSignedWitnessSetCache carries no risk, though a
+        // remaining item would.
+        getSignedWitnessSetCache.remove(hash);
```
This looks redundant, it's removing the same item from the cache again. `hash` is calculated as `new P2PDataStorage.ByteArray(signedWitness.getAccountAgeWitnessHash())` when following the calls to signedWitness.

chimp1984 replied:
I have not looked closely but I took it from the getSignedWitnessSet methods where `signedWitness.getHashAsByteArray()` got compared with `new P2PDataStorage.ByteArray(signedWitness.getAccountAgeWitnessHash())`. Have not looked deeper into the domain if those 2 data are the same actually. If so, yes that second remove is pointless.

sqrrm commented on core/src/main/java/bisq/core/account/sign/SignedWitnessService.java:
```diff
@@ -501,7 +533,22 @@ private void publishSignedWitness(SignedWitness signedWitness) {
+        // Not sure if that is needed as well, tests did succeed in both cases, but seems to be more safe to remove
+        // potential entries with hash as well. A removed item in getSignedWitnessSetCache carries no risk, though a
+        // remaining item would.
+        getSignedWitnessSetCache.remove(hash);
```
This seems redundant, same reason as earlier comment

sqrrm commented:
@chimp1984 I'm waiting for your fixes before merging

chimp1984 committed:
Rename getSignedWitnessSetCache to signedWitnessSetCache

chimp1984 committed:
Rename signedWitnessSetCache to signedWitnessSetByAccountAgeWitnessHash
Change comments (after understanding the domain usage better)
Maintain the set at add and remove methods
Use HashMap instead of ConcurrentHashMap as that class is only used
from UserThread and other fields are not threadsafe either.

chimp1984 committed:
We do not need to lookup in getSignedWitnessMapValues
anymore as we maintain the signedWitnessSetByAccountAgeWitnessHash
at each add/remove operation.

chimp1984 committed:
Use getSignedWitnessSet instead of
iterations over getSignedWitnessMapValues

chimp1984 committed:
Improve arbitratorSignOrphanWitness method

sqrrm reviewed (approved):
utACK

@chimp1984 did you want to add anything else to this PR?
